python/ray/data/block.py

```
@@ -370,7 +370,26 @@ def batch_to_block(batch: DataBatch) -> Block:
 
                 # TODO(ekl) once we support Python objects within Arrow blocks, we
                 # don't need this fallback path.
-                return pd.DataFrame(dict(batch))
+
+                if not isinstance(batch, dict):
+                    return pd.DataFrame(dict(batch))
+                else:
+                    # batch is type of dict[str, np.ndarray]
+                    # check for arrays with ndim > 1
+                    if all(val.ndim == 1 for val in batch.values()):
+                        return pd.DataFrame(batch)
+
+                    # Flatten any ndarray with ndim > 1
+                    pd_batch = dict()
```

We can also just raise `ValueError` on unsupported `DataBatch`.